### PR TITLE
Migrate app-main to Universal API #1420

### PR DESCRIPTION
**Pull request by anatol-sialitski**

### CODE REVIEW

File: src/main/resources/assets/js/launcher/main.ts [hunk 1]
```diff
@@ -235,6 +230,20 @@ class Launcher {
         return `theme-${ThemeManager.getTheme(this.config['theme'])}`;
     };
 
+    private getLauncherUrl(): string {
+        const pathName = window.location.pathname;
+        const adminBaseUrl = this.config['adminBaseUrl'];
+
+        return pathName === `${adminBaseUrl}` || pathName === `${adminBaseUrl}/`
+                  ? UriHelper.joinPath(adminBaseUrl, `/${homeAppToolKey}/home/_/${homeAppToolKey}/launcher`)
```
Comment (rymsha) [inline src/main/resources/assets/js/launcher/main.ts:238]: in theory access to `/com.enonic.xp.app.main/home/_/com.enonic.xp.app.main/launcher` should work even on dashboard

File: src/main/resources/assets/js/launcher/main.ts [hunk 2]
```diff
@@ -235,6 +230,20 @@ class Launcher {
...
+                  : UriHelper.joinPath(pathName, `/_/${homeAppToolKey}/launcher`);
+    }
+
+    private isHomeApp(): boolean {
```
Comment (rymsha) [inline src/main/resources/assets/js/launcher/main.ts:242]: what is so special about home app so it needs to be customized in almost every method?

File: src/main/resources/assets/js/launcher/main.ts (hunk 1 above)
Comment (anatol-sialitski) [reply src/main/resources/assets/js/launcher/main.ts:238]: homeAppToolKey = com.enonic.xp.app.main

**Review by rymsha — APPROVED** [review]

File: src/main/resources/assets/js/launcher/main.ts [hunk 3]
```diff
@@ -585,19 +587,27 @@ const getConfigAttribute = (attribute: string): string => {
     return currentScript?.getAttribute(`data-config-${attribute}`);
 };
 
+const required = (value, msg): string => {
+    if (value === undefined || value === null) {
+        throw msg;
+    }
+    return value;
+};
+
```
Comment (alansemenov) [inline src/main/resources/assets/js/launcher/main.ts:596]:
```Typescript
const getRequiredAttribute = (attrName): string => {
  return required(getConfigAttribute(attrName), `"data-config-${attrName}" is not defined`);
}
```

File: src/main/resources/assets/js/launcher/main.ts [hunk 4]
```diff
@@ -585,19 +587,27 @@ const getConfigAttribute = (attribute: string): string => {
...
 const getConfig = (): JSONObject => {
     return {
         theme: getConfigAttribute('theme') || 'light',
         autoOpenLauncher: getConfigAttribute('auto-open'),
         container: getConfigAttribute('container'),
         customCls: getConfigAttribute('custom-class'),
+        toolBaseUrl: required(getConfigAttribute('tool-base-url'), '"data-config-tool-base-url" is not defined'),
+        appName: required(getConfigAttribute('app-name'), '"data-config-app-name" is not defined'),
```
Comment (alansemenov) [inline src/main/resources/assets/js/launcher/main.ts:604]:
```suggestion
        toolBaseUrl: getRequiredAttribute('tool-base-url'),
        appName: getRequiredAttribute('app-name'),
```

File: src/main/resources/apis/launcher/launcher.html [hunk 5]
```diff
@@ -7,8 +7,8 @@
 
     <div class="launcher-app-container">
         <div class="scrollable-content">
-            <a href="{{homeTool.url}}" target="_blank" data-id="home">
-                <div class="app-row" id="home">
+            <a href="{{homeTool.url}}" target="_blank" data-id="com.enonic.xp.app.main">
+                <div class="app-row" id="com.enonic.xp.app.main">
```
Comment (alansemenov) [inline src/main/resources/apis/launcher/launcher.html:11]: `com.enonic.xp.app.main` should come from the controller.

File: src/main/resources/admin/tools/home/home.js [hunk 6]
```diff
@@ -28,7 +30,13 @@ const generateParams = () => {
         assetsUri: portal.assetUrl({path: ''}),
         launcherPath: admin.getLauncherPath(),
         theme: 'dark',
-        configServiceUrl: portal.serviceUrl({service: 'config'})
+        configServiceUrl: portal.url({
+            path: `/admin/${app.name}/home/_/${app.name}/config`
+        }),
+        toolBaseUrl: portal.url({
+            path: `/admin/${app.name}/home`,
+        }),
```
Comment (alansemenov) [inline src/main/resources/admin/tools/home/home.js:38]: Avoid repeating `/admin/${app.name}/home`

**Review by alansemenov — CHANGES_REQUESTED** [review]

**Review by alansemenov — APPROVED** [review]

**Review by alansemenov — APPROVED** [review]

**Pull request merged**